src/chaboche.jl

```diff
@@ -1,96 +1,167 @@
+    variables::ChabocheVariableState{T} = ChabocheVariableState{T}()
+    variables_new::ChabocheVariableState{T} = ChabocheVariableState{T}()
+    parameters::ChabocheParameterState{T} = ChabocheParameterState{T}()
+    dparameters::ChabocheParameterState{T} = ChabocheParameterState{T}()
+    options::ChabocheOptions = ChabocheOptions()
 end
 
-@with_kw mutable struct Chaboche <: AbstractMaterial
-    drivers :: ChabocheDriverState = ChabocheDriverState()
-    ddrivers :: ChabocheDriverState = ChabocheDriverState()
-    variables :: ChabocheVariableState = ChabocheVariableState()
-    variables_new :: ChabocheVariableState = ChabocheVariableState()
-    parameters :: ChabocheParameterState = ChabocheParameterState()
-    dparameters :: ChabocheParameterState = ChabocheParameterState()
+"""
+    state_to_vector(sigma::U, R::T, X1::U, X2::U) where U <: Symm2{T} where T <: Real
+
+Adaptor for `nlsolve`. Marshal the problem state into a `Vector`.
+"""
+function state_to_vector(sigma::U, R::T, X1::U, X2::U) where U <: Symm2{T} where T <: Real
+    return [tovoigt(sigma); R; tovoigt(X1); tovoigt(X2)]::Vector{T}
 end
 
-function integrate_material!(material::Chaboche)
+"""
+    state_from_vector(x::AbstractVector{<:Real})
+
+Adaptor for `nlsolve`. Unmarshal the problem state from a `Vector`.
+"""
+function state_from_vector(x::AbstractVector{T}) where T <: Real
+    sigma::Symm2{T} = fromvoigt(Symm2{T}, @view x[1:6])
+    R::T = x[7]
+    X1::Symm2{T} = fromvoigt(Symm2{T}, @view x[8:13])
+    X2::Symm2{T} = fromvoigt(Symm2{T}, @view x[14:19])
+    return sigma, R, X1, X2
+end
+
+"""
+    integrate_material!(material::Chaboche{T}) where T <: Real
+
+Chaboche material with two backstresses. Both kinematic and isotropic hardening.
+"""
+function integrate_material!(material::Chaboche{T}) where T <: Real
     p = material.parameters
     v = material.variables
     dd = material.ddrivers
     d = material.drivers
     @unpack E, nu, R0, Kn, nn, C1, D1, C2, D2, Q, b = p
-    mu = E/(2.0*(1.0+nu))
-    lambda = E*nu/((1.0+nu)*(1.0-2.0*nu))
+    lambda, mu = lame(E, nu)
 
     @unpack strain, time = d
     dstrain = dd.strain
     dtime = dd.time
-    @unpack stress, X1, X2, plastic_strain, cumeq, R, jacobian = v
+    @unpack stress, X1, X2, plastic_strain, cumeq, R = v
 
-    jacobian = isotropic_elasticity_tensor(lambda, mu)
+    # elastic part
+    jacobian = isotropic_elasticity_tensor(lambda, mu)  # dσ/dε, i.e. ∂σij/∂εkl
+    stress += dcontract(jacobian, dstrain)  # add the elastic stress increment, get the elastic trial stress
 
-    stress += dcontract(jacobian, dstrain)
-    seff = stress - X1 - X2
-    seff_dev = dev(seff)
-    f = sqrt(1.5)*norm(seff_dev) - (R0 + R)
+    # resulting deviatoric plastic stress (accounting for backstresses Xm)
+    seff_dev = dev(stress - X1 - X2)
+    # von Mises yield function
+    f = sqrt(1.5)*norm(seff_dev) - (R0 + R)  # using elastic trial problem state
+    isplastic = (f > 0.0)
```

Comment:
did you plan to use this on the next line or is it a leftover from your debugging states?

Reply:
Ah, left over from debugging. Fixed. Thanks for the catch.